pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,539 @@
+	}).AsSelectorPreValidated()
+	ipAddresses, err := r.ipAddressLister.List(ipLabelSelector)
+	if err != nil {
+		return fmt.Errorf("unable to refresh the IPAddress block: %v", err)
+	}
+	// Check every IPAddress matches the corresponding Service, and rebuild the state as we think it should be.
+	for _, ipAddress := range ipAddresses {
+		key, err := cache.MetaNamespaceKeyFunc(ipAddress)
+		if err != nil {
+			return err
+		}
+		err = r.syncIPAddress(key)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *RepairIPAddress) svcWorker() {
+	for r.processNextWorkSvc() {
+	}
+}
+
+func (r *RepairIPAddress) processNextWorkSvc() bool {
+	eKey, quit := r.svcQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.svcQueue.Done(eKey)
+
+	err := r.syncService(eKey.(string))
+	r.handleSvcErr(err, eKey)
+
+	return true
+}
+
+func (r *RepairIPAddress) handleSvcErr(err error, key interface{}) {
+	if err == nil {
+		r.svcQueue.Forget(key)
+		return
+	}
+
+	if r.svcQueue.NumRequeues(key) < maxRetries {
+		klog.V(2).InfoS("Error syncing Service, retrying", "service", key, "err", err)
+		r.svcQueue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping Service %q out of the queue: %v", key, err)
+	r.svcQueue.Forget(key)
+	runtime.HandleError(err)
+}
+
+// syncServices reconcile the Service ClusterIPs to verify that each one has the corresponding IPAddress object associated
+func (r *RepairIPAddress) syncService(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	svc, err := r.serviceLister.Services(namespace).Get(name)
+	if err != nil {
+		// nothing to do
+		return nil
+	}
+	if !helper.IsServiceIPSet(svc) {
+		// didn't need a ClusterIP
+		return nil
+	}
+
+	for _, clusterIP := range svc.Spec.ClusterIPs {
+		ip := netutils.ParseIPSloppy(clusterIP)
+		if ip == nil {
+			// ClusterIP is corrupt, ClusterIPs are already validated, but double checking here
+			// in case there are some inconsistencies with the parsers
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s is not a valid IP; please recreate Service", ip)
+			runtime.HandleError(fmt.Errorf("the ClusterIP %s for Service %s/%s is not a valid IP; please recreate Service", ip, svc.Name, svc.Namespace))
+			continue
+		}
+
+		family := getFamilyByIP(ip)
+		if network, ok := r.networkByFamily[family]; !ok {
+			// this service is using an IPFamily no longer configured on cluster
+			r.recorder.Eventf(svc, nil, v1.EventTypeWarning, "ClusterIPNotValid", "ClusterIPValidation", "Cluster IP %s(%s) is of ip family that is no longer configured on cluster; please recreate Service", ip, family)
+			runtime.HandleError(fmt.Errorf("the ClusterIP [%v]: %s for Service %s/%s is of ip family that is no longer configured on cluster; please recreate Service", family, ip, svc.Name, svc.Namespace))
+			continue
+		} else if !network.Contains(ip) {
```

Comment:
You don't need else-if here, logic reset and starts on top due to the `continue` at L339

Reply:
indeed